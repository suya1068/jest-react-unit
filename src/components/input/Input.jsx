import React, { createElement, Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

/**
 * - 기능 설명
 * 1. 사용자가 내용을 입력한다.
 * - properties
 * : message
 * : width(number = default value)
 * : minLength(number = default value)
 * : maxLength(number = default value)
 * : type(types = default value("text"))
 *   - "text" || "number" || "date"
 * : theme
 * : title
 * : description
 * : required
 * - method
 * : setMessage(message)
 * : getMessage()
 * : validCheck() - 예정
 * : getVaildTypes() - 예정
 * : getFormat() - 예정
 * : setFormat() - 예정
 * - EVENT
 * : onKeyDown()
 * : onChange()
 */

export default class Input extends Component {
    constructor(props) {
        super(props);
    }

    render() {
        const { children, minLength, maxLength, message, theme, title, description, required, width, ...rest } = this.props;
        const inputProps = {
            ...rest,
            className: classNames("input-container")
        };

        if (width) {
            const sheet = {};
            // const sheet = document.createElement("style");
            sheet.width = width;
            inputProps.style = sheet;
        }

        return React.createElement("input", inputProps, children);
    }
};

Input.types = {
    text: "text",
    number: "number",
    date: "date"
};

Input.propTypes = {
    message: PropTypes.string,
    width: PropTypes.number,
    minLength: PropTypes.number,
    maxLength: PropTypes.number,
    type: PropTypes.oneOf(Object.keys(Input.types)),
    theme: PropTypes.string,
    title: PropTypes.string,
    description: PropTypes.string,
    required: PropTypes.oneOf(["Y", "N"]),
    placeholder: PropTypes.string
};

Input.defaultProps = {
    message: "",
    width: 100,
    minLength: 0,
    maxLength: 0,
    type: "text",
    theme: "",
    title: "",
    description: "",
    required: "N",
    placeholder: ""
};
