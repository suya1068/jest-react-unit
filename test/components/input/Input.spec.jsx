import React from "react";
import Input from "src/components/input/Input";

describe("# Input Component", () => {
    it("렌더링 가능", () => {
        // shallow는 현재 노드를 렌더링하고 그 주변에 얕은 래퍼를 반환합니다.
        // 단일 노드의 래퍼에서만 호출 할 수 있습니다.
        // shallow([options]) => shallowWrapper
        const input = shallow(<Input/>);
        expect(input).toMatchSnapshot();
    });

    it("width를 설정 할 수 있다.", () => {
        const input = shallow(<Input width={30}/>);
        expect(input).toMatchSnapshot();
    });

    it("placeholder 값을 설정 할 수 있다.", () => {
        const input = shallow(<Input placeholder="테스트1"/>);
        expect(input).toMatchSnapshot();
    });

    it("input type을 설정할 수 있다.", () => {
        const input = shallow(<Input type="text"/>);
        expect(input).toMatchSnapshot();
    });
});

/**
 * 1. 컴포넌트 생성
 * 2. input창의 너비(width)를 설정할 수 있다.
 *      - 입력받거나
 *      - 기본값으로 세팅
 * 3. placeholder 의 값을 설정할 수 있다.
 * 4. type을 설정할 수 있다.
 *      - text || number || date
 * 5. type의 상세 설정이 가능하다 ... (pattern입력) ...생각만 하고 있음.
 */

