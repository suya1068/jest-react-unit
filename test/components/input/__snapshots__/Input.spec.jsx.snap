// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# Input Component input type을 설정할 수 있다. 1`] = `
<input
  className="input-container"
  placeholder=""
  style={
    Object {
      "width": 100,
    }
  }
  type="text"
/>
`;

exports[`# Input Component placeholder 값을 설정 할 수 있다. 1`] = `
<input
  className="input-container"
  placeholder="테스트1"
  style={
    Object {
      "width": 100,
    }
  }
  type="text"
/>
`;

exports[`# Input Component width를 설정 할 수 있다. 1`] = `
<input
  className="input-container"
  placeholder=""
  style={
    Object {
      "width": 30,
    }
  }
  type="text"
/>
`;

exports[`# Input Component 렌더링 가능 1`] = `
<input
  className="input-container"
  placeholder=""
  style={
    Object {
      "width": 100,
    }
  }
  type="text"
/>
`;
